### DISCUSSION TITLE
OSS-Fuzz integration

### DISCUSSION THREAD

**Discussion by guidovranken** [transmission/transmission#3703] [Ideas]
OSS-Fuzz: Continuous Fuzzing for Open Source Software
https://github.com/google/oss-fuzz

I have a few fuzzers ready to be integrated into this repository. These can be run on OSS-Fuzz.

Requisites:

- A Google account (to be able to log in to the dashboard at oss-fuzz.com)
- Willingness to fix the issues that the fuzzer finds

Issues found by the fuzzer are private for 3 months and then they are automatically publicly disclosed.

I've written many fuzzers for OSS-Fuzz and for each integration Google gives a $1,000 USD integration reward. I usually keep this myself but because I'm very partial to the BitTorrent ecosystem I'd like to give the Transmission team the opportunity to collect this reward instead.

Is there any interest in this? I will set up the integration, you don't need to do anything except accept a PR which integrates the fuzzers, and address the bugs found on OSS-Fuzz.

**Comment by ckerr** [comment]
Hi @guidovranken,

The short answer is yes, I'm always interested in making Transmission more stable, including bugs found by fuzzers. So I'd like to learn more & would be interested in a PR to integrate.

I haven't used OSS-Fuzz before and am not up on the details. For example who owns the account that sees the issues for the three month period before they're disclosed?

**Comment by guidovranken** [comment]
Hi @ckerr

Typically only the software maintainers have access to this. In this case that would presumably be you, and any number of other maintainers selected by you. I personally don't need access; I can arrange the integration without ever having access to private bugs.

OSS-Fuzz can run four different builds: one with AddressSanitizer, one with UndefinedBehaviorSanitizer, one with MemorySanitizer and one with AddressSanitizer on 32 bit.

Every day the fuzzers are automatically built from the latest source code, and the OSS-Fuzz system will automatically detect when a previously observed issue has been fixed. Additionally a (public) HTML coverage report is generated daily, like this one for OpenSSL: https://storage.googleapis.com/oss-fuzz-coverage/openssl/reports/20220823/linux/report.html

Using OSS-Fuzz does not cost anything.

**Comment by ckerr** [comment]
Sounds good to me.

Let's convert from an issue to a discussion; there may be ongoing discussion but this isn't a bug to fix or leave open :slightly_smiling_face:

**Comment by guidovranken** [comment]
I've pushed some code here: https://github.com/guidovranken/transmission/tree/fuzzers

This builds the fuzzers if you run cmake using `cmake -DBUILD_FUZZER=ON ..`

The CMake code was derived from https://github.com/h2o/picotls/blob/master/CMakeLists.txt , let me know if that is a problem license-wise.

**Reply by nevack** [reply]
I cannot see any changes, this branch has no commits ahead of upstream's.
Did you pushed your changes?

**Reply by guidovranken** [reply]
Apologies, the commit didn't work because I don't have `clang-format` (I have `clang-format-14`). You should be able to see the commit now.

**Reply by nevack** [reply]
Thanks, left some comments there. Maybe it's viable to create Draft PR in here?

**Reply by jvoisin** [reply]
I opened one here: https://github.com/transmission/transmission/pull/5434
